### ISSUE TITLE
Enabling Non-Linear Utility Functions - WTP Space

### ISSUE THREAD

**Issue by crforsythe** [arteagac/xlogit#3]
Love seeing this new contribution to choice modeling software! Very impressive work. I was wondering if you had any intention to expand xlogit's ability to allow for nonlinear utility specification, such as WTP space estimation a la Train and Weeks (2005)?

**Comment by arteagac** [comment]
HI @crforsythe, I am so glad you like xlogit!
Yes, it would be awesome to implement non-linear utility specifications in xlogit.  I have not extensively worked with non-linear utility models, but I should be able to implement it with some help. Do you have any suggestions on good resources (e.g.,  implementations and datasets) that I can use as reference for non-linear mixed logit specifications in WTP space? Also, if you have some pseudo-code or basic implementation script for this, I would be glad to provide guidance on how to integrate it into xlogit or I could integrate it myself.

**Comment by crforsythe** [comment]
@arteagac Thanks so much for the quick response! I would recommend first looking at [jhelvy's R implementation logitr](https://github.com/jhelvy/logitr). However, I believe he has only implemented normal and log-normal coefficient distributions for his mixed logit models. Additionally, the Revelt and Train (1998) provide gradient and LL formulations for generic utility forms. So far as datasets, logitr provides datasets that can be used as a benchmark as well as several examples in [Apollo's](http://www.apollochoicemodelling.com) documentation.

I would be happy to help you with the development if you'd like to discuss offline. My email address is cforsyth@andrew.cmu.edu.

**Comment by arteagac** [comment]
Fantastic! Thank you so much for suggesting those resources and for offering your help. I just emailed you.

**Comment by arteagac** [comment]
Hi @jhelvy. I integrated your probability and gradient formulation and the estimation is now around **~40% faster.** It is impressive how your formulation significantly reduces the amount of computations. Nice job!

The only little issue I have is getting it to work for panel data. Could you please explain to me how to interpret the term highlighted in red below? I coded it as the product (across panels) of the probabilities for the chosen alternatives but it seems it is not correct.
![image](https://user-images.githubusercontent.com/13966395/168408153-eb584688-e269-47e7-be07-6b928051ce3d.png)

**Comment by jhelvy** [comment]
Wow that was a fast implementation - 40% for a day's work isn't too shabby!

One other thing I forgot to mention that got me a huge speed boost was to do my iteration over the betas instead of the draws for the simulated log-likelihood. That is, when I pre-compute the (X_nt - X_n*), I also include the "standard" draws in that pre-computation and set everything up for each beta. Then later when I'm computing the LL probabilities, I don't have to write a loop to iterate over R (the draws) but rather I iterate over the number of betas. So my code looks a little strange as it doesn't exactly line up with how the equations are written. You can see how I do my pre-computes [here](https://github.com/jhelvy/logitr/blob/master/R/modelInputs.R#L375)

As for working out this panel data part, I remember this one being particularly tricky. Looking through my code for this isn't immediately helpful as I reorganized a lot of the calculations to pre-calculate as much as I could. The relevant parts in my code are [here](https://github.com/jhelvy/logitr/blob/master/R/logit.R#L115) and [here](https://github.com/jhelvy/logitr/blob/master/R/logit.R#L222). For the first part, `logitDrawsPanel <- exp(rowsum(log(logitDraws), d$panelID))`, this is a faster way of computing what you've highlighted in red. `rowsum()` is a fast function in R that allows you to do grouped sums by row, so I'm summing across rows and grouping by the panelID. If I take the log first, then do those row sums, then exponentiate that, it's the same as doing a grouped multiplication. The second code chunk I linked to is where I compute the gradients differently based on the panel structure or not. Sorry this may not be super helpful, but I'm on the road. I'll get back to this Monday if I can.

**Comment by arteagac** [comment]
Yes, I am quite happy with the 40% speed gains. Thank you for sharing your valuable formulation to make it faster.

In xlogit, I do not iterate over individuals, betas, or draws. Instead, I express everything as multidimensional array operations and then I figure out the lowest cost contraction for common matrix operations. I think this helps speed up the operations.

Thank you for the explanation for the panel data part. I will try to make some room this week to implement it. I will let you know how it goes. After fully integrating your formulation, it should be relatively easy to add the WTP functionality.

**Comment by jhelvy** [comment]
Fantastic. Btw, I will eventually add the math from the overleaf doc to the logitr documentation. I'm not sure if there's a convenient way in Python to mostly automate package documentation, but in R there's a great package called [pkgdown](https://pkgdown.r-lib.org/) that does most of the work for you. If there's something like this for Python that would be a great benefit to add and further publicize the package. If I get time this summer I might try and look it up myself as I'd like to learn more about Python packages.

**Comment by arteagac** [comment]
Yes! In Python, I use Sphinx to automatically generate the documentation. Also, I use Jupyter notebooks to create examples, and I further convert these to web page using Sphinx. For instance, see xlogit's documentation website below:

https://xlogit.readthedocs.io/en/latest/notebooks/mixed_logit_model.html

**Comment by jhelvy** [comment]
> 

Perfect. You should put that link in the [main page](https://github.com/arteagac/xlogit) under the "about" section!

**Comment by jhelvy** [comment]
Oh I meant under the main "about" section on the github page, not the readme, like how I have it on the logitr page:

<img width="1244" alt="logitr" src="https://user-images.githubusercontent.com/9012994/168658375-cbda19dd-6da8-496f-b7a8-0dbdde21a172.png">

**Comment by arteagac** [comment]
Oh I see. Yes, it is a great idea. I just added it. Thanks for the suggestion.

**Comment by arteagac** [comment]
Hi @jhelvy, I managed to get the panel data part to work. Now I am having a little challenge estimating fixed (non-random) coefficients, so I figured I would ask in case you can help me figure it out faster.  For instance, assume that in the Yogurt dataset, I want `feat` and `brand` as random coefficients, but I want the `price` as a fixed coefficient. Which of the following gradient functions would you use for `price`? 

![image](https://user-images.githubusercontent.com/13966395/169877461-d1e174be-d2ab-47bd-ada0-91a0596045b6.png)


I assume I have to use the gradient for MixedLogit, because even though the `price` coefficient is not random, the choice probability $P_n*$ involves computation across random draws. Is it right? I tried implementing it this way but the estimation fails to converge. Do you have any additional comments or recommendations regarding this?

**Comment by jhelvy** [comment]
Yes you have to use the mixed logit gradient. What I do in logitr is set up my "standard" draws (halton draws by default) to just be all 0s for the fixed parameters. When I create the draws of the beta coefficients in mixed logit, I shift the standard draws like this: draw * sigma + mu. So if the draw is 0 for fixed parameters, you end up with only the mu parameter (the fixed parameter). [Here's](https://github.com/jhelvy/logitr/blob/master/R/draws.R#L6) the relevant function for shifting the draws. Does this make sense?

**Comment by arteagac** [comment]
I see. Does this mean that you use the same gradient computations for fixed and random coefficients with the only difference being that you set the random draws as zero for fixed coefficients? I tried this approach too, but I am still struggling to reach convergence. Perhaps it is something else in my code. 
The fixed coefficients part is supposed to be the easiest one, but it is giving me the hardest time :sweat_smile:. I will get back to you once I figure it out.

**Comment by jhelvy** [comment]
Yes, if there are any random parameters, then I use the mixed logit gradient for all parameters with a zero for the draws for the fixed coefficients. The math should work out. Like, if you had 100 draws for each of the parameters, then you'll end up with 100 of the same number each of the fixed parameters (just the mu term, no sigma), and 100 different numbers from different distributions for the random parameters (draw*sigma + mu). It's probably a small bug somewhere because I know this approach works. Most of my bugs come from messing up the scaling somewhere (I scale all the data before doing the optimization). You could also try computing the gradient numerically to see if your analytic gradient is producing the same result to try and narrow in on the source of the issue?

**Comment by arteagac** [comment]
Yes, I am actually surprised too because the math works and makes total sense, so I also suspect it's some weird bug. Yes, by looking at the gradients, I can see that those for the fixed coefficients are excessively high, so I might be missing some scaling steps. I will keep trying and keep you posted.

**Comment by arteagac** [comment]
I finally figured it out. The issue was because, while computing the gradient, I was doing the operation $V_{nj} - V_{n*}$ separately for fixed and random coefficients. However, such an operation needs to include both types of coefficients. Now it works like a charm, I will clean up the code and release a new version with the updated formulation in the next few days.

**Comment by jhelvy** [comment]
Woohoo! That's great news! I figured it was a bug somewhere. 

Btw, do you create your draws of the coefficients the same way? As I said before, I create halton draws and then shift them into a normal with draw*sigma + mu, or exp(draw*sigma + mu) for log-normal. I want to add support for other distributions in logitr, but haven't gotten around to it yet. I assume a similar procedure can be used for truncated normal, triangular, etc., but I haven't sat down to figure it out yet. If you've already worked this out it would certainly help me implement it faster in logitr. As this whole exercise shows, there are always different ways to do something, and often times the faster approach is sometimes less intuitive.

**Comment by arteagac** [comment]
Yes, fortunately, it was easy to fix.
Yes, I also create and shift the Halton draws using the `draws*sigma + mu` strategy. Adding extra distributions to logitr should be very easy, as you simply need to use the formulas below to shift the draws to the desired distribution. Just have in mind that this also affects the derivative (gradient), but you can find all the formulas in the [xlogit paper](https://doi.org/10.1016/j.jocm.2021.100339) and I would be more than glad to discuss further or answer any question you may have.

![image](https://user-images.githubusercontent.com/13966395/170689232-69c45b96-0947-4924-898b-7daae015808b.png)

**Comment by crforsythe** [comment]
Woo hoo, @arteagac ! Glad to see it's all working. I look forward to testing out this new feature.

**Comment by arteagac** [comment]
Yes, scipy's algorithm also offers the option of having a single function to return both the objective value and gradient, which reduces computations to a high degree. I will take a look at NLopt for Python in case scipy's function struggles to converge.

I also wanted to ask you and @crforsythe whether you have any suggestions about naming for the parameters and functionality. In a previous conversation with @crforsythe, I mentioned that it would be good to keep the naming in a way that is not hardcoded for WTP models only, but that also works for other similar models (willingness to wait for instance). Therefore, we mentioned that perhaps we could name the `price` input as `scaling_factor`. Also, I was wondering whether we should call this functionality WTP space or if there is a more generic name that may apply to other types of models with similar utility specifications. Do you have any suggestions about the naming of the parameters and the functionality?

**Comment by jhelvy** [comment]
Yes, I agree it might be better to avoid hard-coding in the "willingness to pay" language as it is just one way to scale the utility (though it's by far the most common). "WTP" does have a large body of literature behind it in that it is frequently referred to as WTP, and pretty much every paper that does this sort of scaled utility does so in the "WTP space". But I like calling the argument `scaling_factor` as it does leave the door open to other types of scaling. It must be a continuous variable though, so price and time make sense.

**Comment by arteagac** [comment]
Perfect. I have an additional question. To provide extra flexibility, should we allow to independently use the `scaling_factor` and `lambda` multipliers or these should be tied together? For instance, some people may want to scale their utility function but not include a `lambda` coefficient. At the same time, some people may want a the `lambda` coefficient without scaling the utility. This could be achieved by having two parameters: `scaling_factor` and `lambda_mult` (True or False).

**Comment by jhelvy** [comment]
I'm not sure I understand the question. For a WTP space model, the utility model is by definition scaled by the lambda coefficient: 

$$u_j = \lambda ( \omega X_j - p_j) + \varepsilon_j$$

So whatever you use for $p_j$ is the variable for the `scaling_factor`. For WTP space, you would set the `scaling_factor = "price"`. But you could just set it to some other variable, like `"time"`, and the interpretation of the coefficients would be in units of time instead of currency.

**Comment by arteagac** [comment]
Yes, $p_j$ is the `scaling_factor` for WTP, but for generic models, it can be any continuous variable (e.g. time). My question was whether we should allow to independently include the `scaling_factor` and `lambda` parameters, which enables the estimation of the three scenarios below. I know Scenario 1 is the most common one, but I was just wondering whether we should provide the flexibility for the other two scenarios, as it is very easy to provide such flexibility.

Scenario 1. WTP-like model -> `fit(X, y, ..., scaling_factor="price", lambda_coef=True)`
$$u_j = \lambda(wX_j - p_j) + \epsilon_j$$

Scenario 2. Only scaling -> `fit(X, y, ..., scaling_factor="price", lambda_coef=False)`
$$u_j = wX_j - p_j + \epsilon_j$$

Scenario 3. Only lambda -> `fit(X, y, ..., scaling_factor=None, lambda_coef=True)`
$$u_j = \lambda(wX_j) + \epsilon_j$$

**Comment by arteagac** [comment]
Great. I am glad you mentioned that because I didn't know $\lambda$ could also follow a distribution. For the first version, I will implement it only for a fixed scale parameter, and we can expand it later to enable the estimation of non-fixed lambda.

**Comment by arteagac** [comment]
I finally had the chance to complete the WTP implementation. Sorry for the delay. The code below estimates a Mixed Logit in WTP space for the Yogurt dataset. For 10k draws, the estimation takes 46 seconds in my computer. The estimation results are very similar to logitr's. **Would you guys like to take a look at this functionality (perhaps test an estimation) before I release a new version with this new feature?** To install the current version in the repository you can use:
```
pip install git+https://github.com/arteagac/xlogit.git
```

## Estimation
```python
import pandas as pd
import numpy as np
from xlogit import MixedLogit

df = pd.read_csv("examples/data/yogurt_long.csv")
df["brand_yoplait"] = 1*(df["alt"] == "yoplait")
df["brand_hiland"] = 1*(df["alt"] == "hiland")
df["brand_weight"] = 1*(df["alt"] == "weight")
varnames = ["feat", "brand_yoplait", "brand_hiland", "brand_weight"]

ml = MixedLogit()
ml.fit(X=df[varnames],
       y=df["choice"],
       varnames=varnames,
       ids=df["chid"],
       alts=df["alt"],
       randvars={"feat": "n", "brand_yoplait": "n", "brand_hiland": "n", "brand_weight": "n"},
       n_draws=10000,
       scale_factor=df["price"],
       verbose=2)

ml.summary()
```

OUTPUT: 
```
Optimization terminated successfully.
    Message: CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
    Iterations: 92
    Function evaluations: 0
Estimation time= 46.0 seconds
---------------------------------------------------------------------------
Coefficient              Estimate      Std.Err.         z-val         P>|z|
---------------------------------------------------------------------------
feat                    1.1158147     0.5312314     2.1004307         0.088 .  
brand_yoplait           2.0515495     0.1268223    16.1765715      8.56e-55 ***
brand_hiland           -9.5535598     0.5579497   -17.1226191      6.23e-61 ***
brand_weight           -3.1753436     0.7414522    -4.2826006      8.56e-05 ***
sd.feat                 5.5396861     1.2419099     4.4606185      3.96e-05 ***
sd.brand_yoplait       -0.0011746     1.1310987    -0.0010384         0.798    
sd.brand_hiland        -0.0083729     1.9259843    -0.0043473         0.798    
sd.brand_weight         4.5659533     1.2832545     3.5581042       0.00144 ** 
_lambda                 0.4315093     0.0313476    13.7653006      1.89e-40 ***
---------------------------------------------------------------------------
Significance:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

```

## Prediction
```python
pred_choices, pred_freq = ml.predict(X=df[varnames],
                                     varnames=varnames,
                                     ids=df["chid"],
                                     alts=df["alt"],
                                     scale_factor=df["price"],
                                     n_draws=10000,
                                     return_freq=True)

print("Predicted shares:")
print(pred_freq)

print("Actual shares:")
alt_list, counts = np.unique(df[df["choice"] == 1]["alt"], return_counts=True)
freq = dict(zip(list(alt_list), list(np.round(counts/np.sum(counts), 3))))
print(freq)
```
OUTPUT:
```
Predicted shares:
{'dannon': 0.762, 'weight': 0.029, 'yoplait': 0.209}
Actual shares:
{'dannon': 0.402, 'hiland': 0.029, 'weight': 0.229, 'yoplait': 0.339}
```

**Comment by crforsythe** [comment]
Cristian,

How exciting! I’ll take a look tomorrow and get back to you. Excited to test it out!

Best,
Connor
—
Connor Forsythe
Ph.D. Student, Mechanical Engineering
Carnegie Mellon University
Pronouns: He/Him/His

**Comment by jhelvy** [comment]
This is awesome! I like the argument names too here. Just including the `scale_factor` is all that is truly needed to understand that you're using a WTP space model, and it also gives more flexibility if you want to specify a different scaling factor (e.g. time). In logitr, you also have to specify `modelSpace = 'wtp'`, which is actually rather redundant and forces you into thinking that you can only scale by price. I'm considering whether I should just remove that argument and change `price` to `scaleFactor` as well.

I haven't gotten to try it out yet, but I'll add a few suggestions:

- Run a check internally that whatever string is provided as `scale_factor` isn't also in `varnames` or `randvars`. I very often forget to remove `"price"` from one or both when I'm switching between spaces, and logitr throws an error reminding me to do that.
- I know you haven't added this yet, but you can make the `scale_factor` random as well, which I think should be specified with a separate argument like `rand_scale` or something (in logitr it's `randPrice`). Keeping it separate is related to my first bullet - it's very easy to forget that you had `price` set to a random parameter in `randvars` if you're switching spaces, so if you make it a separate argument you have to be more explicit about whether you want a fixed or random scaled parameter.
- In the summary print out, I suggest renaming `_lambda` to `scale_factor`. I use `lambda` in logitr because I don't want to use `price` (because it's not quite the same as price), but since you're not calling it "price" you could use `scale_factor` as the parameter name. I've never liked using greek symbol names in the print outs as people use different symbols for different things, but I'm still using `lambda` since I used `price` as the scale argument name. This is more motivation for me to switch the naming in logitr.

**Comment by arteagac** [comment]
Hi @jhelvy , thanks a lot for your suggestions. See my comments below.

> - Run a check internally that whatever string is provided as `scale_factor` isn't also in `varnames` or `randvars`. I very often forget to remove `"price"` from one or both when I'm switching between spaces, and logitr throws an error reminding me to do that.

Yes, I need to find a way to validate that the variable used as `scale_factor` is not present in `varnames`. The challenge is that `xlogit` takes as input pure `numpy` arrays, so it does not know the column names from the `pandas` dataframe. I think I can find a workaround for this.

> - I know you haven't added this yet, but you can make the `scale_factor` random as well, which I think should be specified with a separate argument like `rand_scale` or something (in logitr it's `randPrice`). Keeping it separate is related to my first bullet - it's very easy to forget that you had `price` set to a random parameter in `randvars` if you're switching spaces, so if you make it a separate argument you have to be more explicit about whether you want a fixed or random scaled parameter.

Sure, once we implement this feature, I will include a new `rand_scale` argument to specify the desired distribution for `scale_factor`. 
> - In the summary print out, I suggest renaming `_lambda` to `scale_factor`. I use `lambda` in logitr because I don't want to use `price` (because it's not quite the same as price), but since you're not calling it "price" you could use `scale_factor` as the parameter name. I've never liked using greek symbol names in the print outs as people use different symbols for different things, but I'm still using `lambda` since I used `price` as the scale argument name. This is more motivation for me to switch the naming in logitr.

Sure, I will rename the `_lambda` parameter as `scale_factor`. I agree it makes more sense than using greek symbol's names.
